// xgplayer-music/es/preset.js
import { createClass as _createClass, classCallCheck as _classCallCheck, objectSpread2 as _objectSpread2 } from "./_virtual/_rollupPluginBabelHelpers.js";
import { Util } from "xgplayer";
import Music from "./music.js";
import { MusicCover } from "./plugins/musicCover.js";
import { MusicBackward } from "./plugins/musicBackward.js";
import { MusicPrev } from "./plugins/musicPrev.js";
import { MusicForward } from "./plugins/musicForward.js";
import { MusicNext } from "./plugins/musicNext.js";
import { MusicMeta } from "./plugins/musicMeta.js";
function initMusicConfig(playerConfig) {
  if (Util.typeOf(playerConfig.controls) !== "Object") {
    playerConfig.controls = {
      mode: "flex",
      initShow: true
    };
  } else {
    var _c = playerConfig.controls;
    playerConfig.controls = _objectSpread2({
      mode: "flex",
      initShow: true
    }, _c);
  }
  var _p = playerConfig.play || {};
  playerConfig.play = _objectSpread2({
    index: 3
  }, _p);
  playerConfig.marginControls = true;
  playerConfig.mediaType = "audio";
}
var MusicPreset = /* @__PURE__ */ _createClass(function MusicPreset2(options, playerConfig) {
  _classCallCheck(this, MusicPreset2);
  console.log("config", options, playerConfig);
  console.log("playerConfig", Util.typeOf(playerConfig.controls) !== "Object");
  initMusicConfig(playerConfig);
  this.plugins = [
    Music,
    MusicBackward,
    MusicPrev,
    MusicForward,
    MusicNext,
    MusicCover,
    MusicMeta
  ];
  this.ignores = ["start", "cssFullScreen", "fullscreen", "pip", "enter", "replay", "pc", "loading", "poster"];
});
export { MusicPreset as default };
