// xgplayer-music/es/plugins/musicNext.js
import { inherits as _inherits, createSuper as _createSuper, createClass as _createClass, classCallCheck as _classCallCheck, defineProperty as _defineProperty, assertThisInitialized as _assertThisInitialized } from "../_virtual/_rollupPluginBabelHelpers.js";
import { Plugin } from "xgplayer";
var MusicNext = /* @__PURE__ */ function(_Plugin) {
  _inherits(MusicNext2, _Plugin);
  var _super = _createSuper(MusicNext2);
  function MusicNext2() {
    var _this;
    _classCallCheck(this, MusicNext2);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    _defineProperty(_assertThisInitialized(_this), "onClick", function(e) {
      var _assertThisInitialize = _assertThisInitialized(_this), player = _assertThisInitialize.player;
      e.preventDefault();
      e.stopPropagation();
      player.plugins.music.next();
    });
    return _this;
  }
  _createClass(MusicNext2, [{
    key: "afterCreate",
    value: function afterCreate() {
      var _this2 = this;
      this.initIcons();
      var ev = ["click", "touchstart"];
      ev.forEach(function(item) {
        _this2.bind(item, _this2.onClick);
      });
    }
  }, {
    key: "registerIcons",
    value: function registerIcons() {
      return {
        musicNext: {
          icon: ' <svg xmlns="http://www.w3.org/2000/svg" width="40" height="32" viewBox="100 200 1024 1024">\n        <path d="M800 380v768h-128v-352l-320 320v-704l320 320v-352z">\n      </svg>',
          class: "xg-icon-mnext"
        }
      };
    }
  }, {
    key: "initIcons",
    value: function initIcons() {
      var icons = this.icons;
      this.appendChild(".xgplayer-icon", icons.musicNext);
    }
  }, {
    key: "destroy",
    value: function destroy() {
      var _this3 = this;
      var ev = ["click", "touchstart"];
      ev.forEach(function(item) {
        _this3.unbind(item, _this3.onClick);
      });
    }
  }, {
    key: "render",
    value: function render() {
      return '<xg-icon class="xgplayer-next">\n            <div class="xgplayer-icon">\n            </div>\n          </xg-icon>';
    }
  }], [{
    key: "pluginName",
    get: function get() {
      return "musicnext";
    }
  }, {
    key: "defaultConfig",
    get: function get() {
      return {
        index: 4,
        position: Plugin.POSITIONS.CONTROLS_LEFT
      };
    }
  }]);
  return MusicNext2;
}(Plugin);
export { MusicNext };
