// xgplayer-music/es/plugins/musicPrev.js
import { inherits as _inherits, createSuper as _createSuper, createClass as _createClass, classCallCheck as _classCallCheck, defineProperty as _defineProperty, assertThisInitialized as _assertThisInitialized } from "../_virtual/_rollupPluginBabelHelpers.js";
import { Plugin } from "xgplayer";
var MusicPrev = /* @__PURE__ */ function(_Plugin) {
  _inherits(MusicPrev2, _Plugin);
  var _super = _createSuper(MusicPrev2);
  function MusicPrev2() {
    var _this;
    _classCallCheck(this, MusicPrev2);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    _defineProperty(_assertThisInitialized(_this), "onClick", function(e) {
      var _assertThisInitialize = _assertThisInitialized(_this), player = _assertThisInitialize.player;
      e.preventDefault();
      e.stopPropagation();
      player.plugins.music.prev();
    });
    return _this;
  }
  _createClass(MusicPrev2, [{
    key: "afterCreate",
    value: function afterCreate() {
      var _this2 = this;
      this.initIcons();
      var ev = ["click", "touchstart"];
      ev.forEach(function(item) {
        _this2.bind(item, _this2.onClick);
      });
    }
  }, {
    key: "registerIcons",
    value: function registerIcons() {
      return {
        musicPrev: {
          icon: ' <svg xmlns="http://www.w3.org/2000/svg" width="40" height="32" viewBox="300 200 1024 1024">\n        <path d="M600 1140v-768h128v352l320-320v704l-320-320v352zz"></path>\n    </svg>',
          class: "xg-icon-mprev"
        }
      };
    }
  }, {
    key: "initIcons",
    value: function initIcons() {
      var icons = this.icons;
      this.appendChild(".xgplayer-icon", icons.musicPrev);
    }
  }, {
    key: "destroy",
    value: function destroy() {
      var _this3 = this;
      var ev = ["click", "touchstart"];
      ev.forEach(function(item) {
        _this3.unbind(item, _this3.onClick);
      });
    }
  }, {
    key: "render",
    value: function render() {
      return '<xg-icon class="xgplayer-next">\n            <div class="xgplayer-icon">\n            </div>\n          </xg-icon>';
    }
  }], [{
    key: "pluginName",
    get: function get() {
      return "musicprev";
    }
  }, {
    key: "defaultConfig",
    get: function get() {
      return {
        index: 2,
        position: Plugin.POSITIONS.CONTROLS_LEFT
      };
    }
  }]);
  return MusicPrev2;
}(Plugin);
export { MusicPrev };
