// xgplayer-music/es/plugins/musicForward.js
import { inherits as _inherits, createSuper as _createSuper, createClass as _createClass, classCallCheck as _classCallCheck, defineProperty as _defineProperty, assertThisInitialized as _assertThisInitialized } from "../_virtual/_rollupPluginBabelHelpers.js";
import { Plugin } from "xgplayer";
var MusicForward = /* @__PURE__ */ function(_Plugin) {
  _inherits(MusicForward2, _Plugin);
  var _super = _createSuper(MusicForward2);
  function MusicForward2() {
    var _this;
    _classCallCheck(this, MusicForward2);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call.apply(_super, [this].concat(args));
    _defineProperty(_assertThisInitialized(_this), "onClick", function(e) {
      var _assertThisInitialize = _assertThisInitialized(_this), player = _assertThisInitialize.player;
      e.preventDefault();
      e.stopPropagation();
      player.plugins.music.forward();
    });
    return _this;
  }
  _createClass(MusicForward2, [{
    key: "afterCreate",
    value: function afterCreate() {
      var _this2 = this;
      this.initIcons();
      var ev = ["click", "touchstart"];
      ev.forEach(function(item) {
        _this2.bind(item, _this2.onClick);
      });
    }
  }, {
    key: "registerIcons",
    value: function registerIcons() {
      return {
        musicForward: {
          icon: '<svg xmlns="http://www.w3.org/2000/svg" width="40" height="40" viewBox="-2 0 21 15">\n        <path d="m 2,2.99996 0,10 7,-5 -7,-5 z m 7,5 0,5 7,-5 -7,-5 0,5 z m 7,0 0,0 z"></path>\n      </svg>',
          class: "xg-icon-mforward"
        }
      };
    }
  }, {
    key: "initIcons",
    value: function initIcons() {
      var icons = this.icons;
      this.appendChild(".xgplayer-icon", icons.musicForward);
    }
  }, {
    key: "destroy",
    value: function destroy() {
      var _this3 = this;
      var ev = ["click", "touchstart"];
      ev.forEach(function(item) {
        _this3.unbind(item, _this3.onClick);
      });
    }
  }, {
    key: "render",
    value: function render() {
      return '<xg-icon class="xgplayer-forward">\n            <div class="xgplayer-icon">\n            </div>\n          </xg-icon>';
    }
  }], [{
    key: "pluginName",
    get: function get() {
      return "musicforward";
    }
  }, {
    key: "defaultConfig",
    get: function get() {
      return {
        index: 5,
        position: Plugin.POSITIONS.CONTROLS_LEFT
      };
    }
  }]);
  return MusicForward2;
}(Plugin);
export { MusicForward };
