// xgplayer-music/es/analyze/render.js
import { objectSpread2 as _objectSpread2 } from "../_virtual/_rollupPluginBabelHelpers.js";
import drawBars from "./drawBars.js";
import drawVertLines from "./drawVertlines.js";
import drawLightning from "./drawLightning.js";
import drawDoubleline from "./drawDoubleline.js";
import drawDoubleBars from "./drawDoubleBars.js";
import drawWaves from "./drawWaves.js";
var modeMap = {
  bars: drawBars,
  vertLines: drawVertLines,
  lightning: drawLightning,
  doubleLine: drawDoubleline,
  doubleBars: drawDoubleBars,
  waves: drawWaves
};
var frameRateMap = {
  bars: 2,
  vertLines: 1,
  lightning: 2,
  doubleLine: 1,
  doubleBars: 2,
  waves: 2
};
var MODES = {
  BARS: "bars",
  VERT_LINES: "vertLines",
  LIGHTNING: "lightning",
  DOUBLE_LINE: "doubleLine",
  DOUBLE_BARS: "doubleBars",
  WAVES: "waves"
};
function render(data, canvas) {
  var options = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : {};
  var frame = arguments.length > 3 ? arguments[3] : void 0;
  if (!canvas) {
    return;
  }
  options = _objectSpread2({}, options);
  if (!options.stroke)
    options.stroke = 1;
  if (!options.colors)
    options.colors = ["#ff8177", "#cf556c", "#f99185", "#b12a5b"];
  var ctx = canvas.getContext("2d");
  var h = canvas.height;
  var w = canvas.width;
  ctx.strokeStyle = options.colors[0];
  ctx.lineWidth = options.stroke;
  var context = {
    data,
    options,
    ctx,
    h,
    w
  };
  if (typeof options.mode === "string")
    options.mode = [options.mode];
  options.mode.forEach(function(mode) {
    if (frame % frameRateMap[mode] === 0) {
      ctx.clearRect(0, 0, w, h);
      ctx.beginPath();
      modeMap[mode](context);
    }
  });
}
export { MODES, render as default };
